### DISCUSSION TITLE
Reproducible global pooling operation?

### DISCUSSION THREAD

**Discussion by yunchong-song** [pyg-team/pytorch_geometric#5205] [Q&A] [answered]
Hi, I use PyG to implement my graph pooling model, however, it seems PyG suffers from numerical unstable on GPUs when using `torch_scatter` based methods, I followed https://github.com/pyg-team/pytorch_geometric/issues/2788#issuecomment-877294792 to fix the `MessagePassing` operation with adjacent matrix in `SparseTensor` format, it works pretty well.

However when I need to implement a graph pooling layer, I need some global pooling operations like `global_mean_pool, global_max_pool, global_add_pool` , I can't find **reproducible** alternatives in PyG, I want to ask if PyG provides such alternatives? If not, how can I make them(`global_mean_pool, global_max_pool, global_add_pool`) **reproducible**?

**Comment by rusty1s** [comment, marked as answer]
You can use `segment_csr` from `torch-scatter`:
```python
from torch_scatter import segment_csr

global_x = segment_csr(x, batch.ptr, reduce='sum')
```

**Reply by yunchong-song** [reply]
Thank you, I'll check this.

**Reply by yunchong-song** [reply]
BTW, is there a time-efficient way to generate `ptr` tensor from ordered `batch` tansor? without for-loop.

**Reply by rusty1s** [reply]
The mini-batch will auto-generate it for you. Alternatively, you can do:
```python
deg = torch_geometric.utils.degree(batch)
ptr = deg.cumsum(dim=0)
ptr = torch.cat([torch.tensor([0]), ptr], dim=0)
```

**Reply by yunchong-song** [reply]
Perfect, thanks!!
